vnext/Desktop/WebSocketResourceFactory.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include "pch.h"
+
+#include <WinRTWebSocketResource.h>
+#include "BeastWebSocketResource.h"
+#include "WinHTTPWebSocketResource.h"
+
+using std::string;
+using std::unique_ptr;
+
+namespace Microsoft::React
+{
+#pragma region IWebSocketResource static members
+
+/*static*/ unique_ptr<IWebSocketResource> IWebSocketResource::Make(const string &urlString) {
+  if (true) //TODO: Feature-gate this.
+  {
+    //return unique_ptr<IWebSocketResource>(new WinHTTPWebSocketResource(urlString));
+    return unique_ptr<IWebSocketResource>(new WinRTWebSocketResource(urlString));
```

Comment:
That wouldn't compile.
The expected return value is `unique_ptr<IWebSocketResource>`.
`make_unique` would have to take `WinRTWebSocketResource` as a type parameter.

Reply:
Did you test this? `make_unique<WinRTWebSocketResource>` should work fine. We'll do implicit conversion when calling the ctor for the base class `unique_ptr`.

Reply:
Yes.
I tested it back when implementing `BeastWebSocketResource`, and also tested it when implementing `WinRTWebSocketResource`.

It still does not compile.

Reply:
std pointer types are covariant, this should work as Nick says

---
In reply to: [394404280](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r394404280) [](ancestors = 394404280)

Reply:
Except int doesn't (I had tried this before):
```
    //return unique_ptr<IWebSocketResource>(new WinRTWebSocketResource(urlString));
    return std::make_unique<WinRTWebSocketResource>(urlString);
```

```
>c:\program files (x86)\microsoft visual studio\2019\enterprise\vc\tools\msvc\14.16.27023\include\memory(2084): error C2248: 'Microsoft::React::WinRTWebSocketResource::~WinRTWebSocketResource': cannot access private member declared in class 'Microsoft::React::WinRTWebSocketResource'
```

Maybe there'es a compiler config that needs to be enable for this purpose?
Is there a specific way the class or interface should be declared?
Else, let's leave it like this for now.

Reply:
The reason it doesn't is because you left the destructor (and the move constructor) as private.
classes have default private visibility for their members, you have to say public: before the thing you want to make public. The error is correct, you should fix it.

---
In reply to: [395916156](https://github.com/microsoft/react-native-windows/pull/4330#discussion_r395916156) [](ancestors = 395916156)

Reply:
> Something to do with the constructor being overriden?

Yeah, I meant destructor.
I'm aware it was kept private. Didn't forsee this issue.

Making int public solves the prublem.
Switching to `make_unique`.